### PR TITLE
add a general pull request template

### PR DESCRIPTION
**Pull request by FFMMM**
### overview

This adds a `pull_request_template.md`. I followed the instructions on  https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository 

For the actual content, I pulled from our internal discussion documented [here](https://docs.google.com/document/d/1dK0eVwE4DfPi2BYaGRGRXglZlpIaNh8wfF3CcP5NaQA/edit#bookmark=id.s97arnxz7h9f). This PR checklist itself is taken from the internal eng handbook.

Something more to consider when reviewing this is: *how does this checklist impact outside contributors?* As in, is it clear enough? Do we need to add more doc pointers? etc.

### future work

I think we can easily start with a pull request template per repo but then graduate to having package specific checklists.

Signed-off-by: FFMMM <FFMMM@users.noreply.github.com>

### CODE REVIEW

**Review by DanStough — APPROVED** [review]
Looks good! 

As a newbie, it might be helpful to call out in the checklist which manual labels should be considered for a PR. Like `backport/X` or `pr/no-changlelog` (thanks for adding that to my PR btw!).

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+CHANGELOG: `no-impact` / `Added` / `Changed` / `Deprecated` / `Removed` / `Fixed` / `Security`
+
+Describe how these changes affect the customer in one line. This is the thing you are going to add to a changelog.txt file.
+
+> CHANGELOG entries will be pulled directly into the release changelog, so use
+customer-facing language. Start with one of the active verbs above,
+write a full sentence, and end with a period. If these changes do not
+impact the customer, then this line should read CHANGELOG: no-impact.
```
Comment (kisunji) [inline .github/pull_request_template.md:8]: This doesn't seem to fit the way we do changelogs. Is it from some other template?

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+* In the case of bugs, describe how to replicate
+* If any manual tests were done, document the steps and the conditions to replicate
+* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding
+
+### Links
+Include any links here that might be helpful for people reviewing your PR (GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.
+
+### PR Checklist
+
+Please fill in this checklist PRIOR to opening a PR.
```
Comment (kisunji) [inline .github/pull_request_template.md:23]: I don't think we need to be prescriptive here; some people add TODOs as they work on an open PR

Comment (kisunji) [inline .github/pull_request_template.md:23]: I don't think we need to be prescriptive here; some people add TODOs as they work on an open PR

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+* [ ] lint
+* [ ] testing
+* [ ] external facing docs updated
+* [ ] not a security concern
+* [ ] checklist folder consulted
```
Comment (kisunji) [inline .github/pull_request_template.md:29]: I didn't know we had this! Is there a link?

Comment (kisunji) [inline .github/pull_request_template.md:29]: I didn't know we had this! Is there a link?

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+### Links
+Include any links here that might be helpful for people reviewing your PR (GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.
+
+### PR Checklist
+
+Please fill in this checklist PRIOR to opening a PR.
+
+* [ ] lint
+* [ ] testing
```
Comment (kisunji) [inline .github/pull_request_template.md:26]: These two will fail at CI so I don't think we need to call them out

Comment (kisunji) [inline .github/pull_request_template.md:26]: These two will fail at CI so I don't think we need to call them out

**Review by kisunji — COMMENTED** [review]
My personal preference would be to keep the template as minimal as possible with self-explanatory Headers and _maybe_ a one-liner description (I don't like having to highlight lines then deleting them but that's me being nit-picky)

Maybe we could have a section like `### Note for reviewers` where you can add meta-information on what files should get attention, whether to view commit-by-commit, hide whitespace, etc.

File: .github/pull_request_template.md (hunk 1 above)
Comment (FFMMM) [reply .github/pull_request_template.md:8]: This is lifted from [here](https://eng-handbook.hashicorp.services/git-and-github/#templates), which AFAIK, is a terraform checklist.

Are you saying that the contents themselves are not applicable to us? I think it could be valuable to follow that pattern for our changelogs too 🤔

File: .github/pull_request_template.md (hunk 4 above)
Comment (FFMMM) [reply .github/pull_request_template.md:26]: hm, I've seen PRs before that don't add these. My own PRs sometimes don't lint 🙈 bc I forget to do it; and I think it's fine to use the CI as a linter (💸 concern maybe?).

---

On the testing front, I have also seen changes go in without a lot of test coverages (certainly the odd PR, not the norm) but I want testing/ quality to be forefront of my mind 

---

I think all the `PR Checklist` items can be re-worded to better describe that ^ 😆

File: .github/pull_request_template.md (hunk 3 above)
Comment (FFMMM) [reply .github/pull_request_template.md:29]: yeah, I can add a link ⌨️

File: .github/pull_request_template.md (hunk 4 above)
Comment (markan) [reply .github/pull_request_template.md:26]: I think lint doesn't belong here, because it's automated and there shouldn't need to be any user action to check off.
As far as testing goes; maybe something like 'any needed test coverage for affected area'

File: .github/pull_request_template.md (hunk 1 above)
Comment (markan) [reply .github/pull_request_template.md:8]: I'm not sure how the no-impact option fits in with the our automated changelog generation tool. It might be worth including the keywords that it looks at for convenience. I believe the pattern is ```release-note: 
followed by one of breaking-change, bug, deprecation, enhancement, feature, improvement, note, security
We historically have been using lowercase, and I'd argue for keeping it that way for consistiency.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+### Description
+Describe why you're making this change, in plain English.
+
+### Testing & Repro steps
+* In the case of bugs, describe how to replicate
+* If any manual tests were done, document the steps and the conditions to replicate
+* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding
+
+### Links
+Include any links here that might be helpful for people reviewing your PR (GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.
```
Comment (markan) [inline .github/pull_request_template.md:19]: I would like to suggest we start including links to internal systems here if they don't leak customer information. E.g. Asana tickets, consul-enterprise issues, etc. While we should transcribe and perhaps summarize that information in the PR, having the links helps keep things from falling through the cracks and can provide context if something is lost in translation.

**Review by markan — CHANGES_REQUESTED** [review]

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+> CHANGELOG entries will be pulled directly into the release changelog, so use
+customer-facing language. Start with one of the active verbs above,
+write a full sentence, and end with a period. If these changes do not
+impact the customer, then this line should read CHANGELOG: no-impact.
+
+### Description
+Describe why you're making this change, in plain English.
+
+### Testing & Repro steps
```
Comment (boxofrad) [inline .github/pull_request_template.md:13]: Tiny nit, but I think we should avoid using abbreviations that new folks might not be familiar with.

```suggestion
### Testing & Reproduction steps
```

File: .github/pull_request_template.md (hunk 5 above)
Comment (boxofrad) [reply .github/pull_request_template.md:19]: Maybe a slightly hair-brained idea, but I wonder if there's a convenient way to use our internal URL shortener (which requires SSO) for this kinda thing? Like a Chrome plugin that spots when you're trying to paste an internal link and replaces it with an obfuscated version.

File: .github/pull_request_template.md (hunk 1 above)
Comment (boxofrad) [reply .github/pull_request_template.md:8]: I don't think this belongs in the PR description, as it's already enforced by a GitHub action and (as others have said) doesn't match the way we actually manage our changelog.

**Review by boxofrad — COMMENTED** [review]
Great job! Thanks for taking this on 🙌🏻

File: .github/pull_request_template.md (hunk 1 above)
Comment (FFMMM) [reply .github/pull_request_template.md]: Thanks everyone for the feedback -- per that, I deleted this section.

File: .github/pull_request_template.md (hunk 4 above)
Comment (FFMMM) [reply .github/pull_request_template.md]: removed per feedback

File: .github/pull_request_template.md (hunk 5 above)
Comment (FFMMM) [reply .github/pull_request_template.md]: I added guidance to internal links and how to use hashi links

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+### Testing & Reproduction steps
+* In the case of bugs, describe how to replicate
+* If any manual tests were done, document the steps and the conditions to replicate
+* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding
+
+### Links
+Include any links here that might be helpful for people reviewing your PR (Asana tasks, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.
+
+Please be mindful not to leak any customer or confidential information. Feel free to use [HashiLinks](https://go.hashi.co/) to obfuscate links as needed.
```
Comment (boxofrad) [inline .github/pull_request_template.md:12]: It's worth considering that the audience for this template is broader than just the Consul team, and also includes external contributors (who won't know what HashiLinks is, for example).

```suggestion
Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.
```

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/650135.
